### ISSUE TITLE
[Bug]: babel fails to process convert-source-map

### ISSUE THREAD

**Issue by cowwoc** [babel/babel#15352]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

@babel/register or @babel/node

### Input code

I seem to be experiencing [Bug #12665](https://github.com/babel/babel/issues/11964) in `@babel/register` version `7.18.9`.

The first time I run gulp the build works. The second time I run it I get `node_modules\convert-source-map\index.js: _convertSourceMap(...).fromObject is not a function`.

### Configuration

```js
const presets =
  [
    "@babel/preset-typescript",
    [
      "@babel/preset-env",
      {
        useBuiltIns: "usage",
        corejs:
          {
            version: "3.27.1",
            proposals: true
          }
        // debug: true
      }
    ]
  ];

const ignore =
  [
    "**/node_modules/**"
  ];

module.exports =
  {
    presets,
    ignore
  };
```

### Current and expected behavior

The first time I run I get:

```shell
"C:\Program Files\nodejs\node.exe" C:\Users\Gili\Documents\requirements.js\.yarn\releases\yarn-3.3.1.cjs run --binaries-only gulp --color --gulpfile C:\Users\Gili\Documents\requirements.js\gulpfile.babel.js default --mode=DEBUG
[14:19:22] Requiring external module @babel/register
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\Gili\Documents\requirements.js\.yarn\cache\typescript-patch-7bc527fc0a-1caaea6cb7.zip\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\Gili\Documents\requirements.js\.yarn\cache\rollup-npm-3.10.0-a133550ae4-31a882689c.zip\node_modules\rollup\dist\shared\rollup.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\Gili\Documents\requirements.js\.yarn\cache\terser-npm-5.16.1-d407c2c364-cb52412350.zip\node_modules\terser\dist\bundle.min.js as it exceeds the max of 500KB.
[14:19:32] DEBUG mode detected
[14:19:33] Using gulpfile ~\Documents\requirements.js\gulpfile.babel.js
[14:19:33] Starting 'default'...
Hello world
[14:19:33] Finished 'default' after 2.17 ms
````

The second time I run I get:

```shell
"C:\Program Files\nodejs\node.exe" C:\Users\Gili\Documents\requirements.js\.yarn\releases\yarn-3.3.1.cjs run --binaries-only gulp --color --gulpfile C:\Users\Gili\Documents\requirements.js\gulpfile.babel.js default --mode=DEBUG
[16:00:46] Requiring external module @babel/register
C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\lib\transformation\index.js:47
    throw e;
    ^

TypeError: C:\Users\Gili\Documents\requirements.js\.yarn\cache\convert-source-map-npm-1.9.0-e294555f4b-dc55a1f28d.zip\node_modules\convert-source-map\index.js: _convertSourceMap(...).fromObject is not a function
    at generateCode (C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\src\transformation\file\generate.ts:70:43)
    at run (C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\src\transformation\index.ts:59:48)
    at run.next (<anonymous>)
    at transform (C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\src\transform.ts:29:20)
    at transform.next (<anonymous>)
    at evaluateSync (C:\Users\Gili\Documents\requirements.js\.yarn\cache\gensync-npm-1.0.0-beta.2-224666d72f-a7437e58c6.zip\node_modules\gensync\index.js:251:28)
    at fn (C:\Users\Gili\Documents\requirements.js\.yarn\cache\gensync-npm-1.0.0-beta.2-224666d72f-a7437e58c6.zip\node_modules\gensync\index.js:89:14)
    at stopHiding - secret - don't use this - v1 (C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\src\errors\rewrite-stack-trace.ts:97:14)
    at Object.transformSync (C:\Users\Gili\Documents\requirements.js\.yarn\cache\@babel-core-npm-7.20.12-1f7c5fb4ea-62e6c3e214.zip\node_modules\@babel\core\src\transform.ts:66:52)
    at exports.transformSync (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\worker\transform.js:94:15)      
    at Function.handleMessage (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\worker\handle-message.js:23:16)
    at LocalClient.<anonymous> (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\worker-client.js:134:86)      
    at LocalClient.transform (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\worker-client.js:64:47)
    at compile (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\hook.js:58:25)
    at compileBabel7 (C:\Users\Gili\Documents\requirements.js\.yarn\__virtual__\@babel-register-virtual-5b77fbcf59\0\cache\@babel-register-npm-7.18.9-bcdce8aed1-4aeaff97e0.zip\node_modules\@babel\register\lib\hook.js:49:14)
    at Module._compile (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:130:29)
    at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
    at require$$0.Module._extensions..js (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19395:33)
    at Object.newLoader [as .js] (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1096:32)
    at Function.require$$0.Module._load (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19232:14)
    at Module.require (node:internal/modules/cjs/loader:1120:19)
    at require (node:internal/modules/helpers:112:18)
    at Object.<anonymous> (C:\Users\Gili\Documents\requirements.js\.yarn\cache\vinyl-sourcemap-npm-1.1.0-e961858bf5-e7174851fa.zip\node_modules\vinyl-sourcemap\lib\/helpers.js:8:15)
    at Module._compile (node:internal/modules/cjs/loader:1239:14)
    at Module._compile (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
    at require$$0.Module._extensions..js (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19395:33)
    at Object.newLoader [as .js] (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1096:32)
    at Function.require$$0.Module._load (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19232:14)
    at Module.require (node:internal/modules/cjs/loader:1120:19)
    at require (node:internal/modules/helpers:112:18)
    at Object.<anonymous> (C:\Users\Gili\Documents\requirements.js\.yarn\cache\vinyl-sourcemap-npm-1.1.0-e961858bf5-e7174851fa.zip\node_modules\vinyl-sourcemap\/index.js:5:15)
    at Module._compile (node:internal/modules/cjs/loader:1239:14)
    at Module._compile (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
    at require$$0.Module._extensions..js (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19395:33)
    at Object.newLoader [as .js] (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1096:32)
    at Function.require$$0.Module._load (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19232:14)
    at Module.require (node:internal/modules/cjs/loader:1120:19)
    at require (node:internal/modules/helpers:112:18)
    at Object.<anonymous> (C:\Users\Gili\Documents\requirements.js\.yarn\cache\vinyl-fs-npm-3.0.3-7379431c37-9483663259.zip\node_modules\vinyl-fs\lib\src\/sourcemap.js:4:17)
    at Module._compile (node:internal/modules/cjs/loader:1239:14)
    at Module._compile (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
    at require$$0.Module._extensions..js (C:\Users\Gili\Documents\requirements.js\.pnp.cjs:19395:33)
    at Object.newLoader [as .js] (C:\Users\Gili\Documents\requirements.js\.yarn\cache\pirates-npm-4.0.5-22f8e827ce-c9994e61b8.zip\node_modules\pirates\lib\index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1096:32) {
  code: 'BABEL_GENERATE_ERROR'
}

Node.js v19.4.0

Process finished with exit code 1
```

Deleting `node_modules` and repeating the process results in the first output.

Expected behavior: No error on second run.

### Environment

  System:
    OS: Windows 10 10.0.19045
  Binaries:
    Node: 19.4.0 - ~\AppData\Local\Temp\xfs-30db7cf1\node.CMD
    Yarn: 3.3.1 - ~\AppData\Local\Temp\xfs-30db7cf1\yarn.CMD
    npm: 9.2.0 - C:\Program Files\nodejs\npm.CMD

**Comment by cowwoc** [comment]
Another detail that is probably related to this is that I'm using yarn 3.3.1.

**Comment by liuxingbaoyu** [comment]
Have you tried turning off yarn pnp?
Also can you provide a reproducible repository?

**Comment by nicolo-ribaudo** [comment]
@cowwoc I tried running the reproduction originally posted in https://github.com/babel/babel/issues/11964 but it still works, do you have an example I could use?

**Comment by cowwoc** [comment]
@nicolo-ribaudo I don't have a minimal testcase, but here is a project I am actively experiencing this problem on:

1. Clone https://github.com/cowwoc/requirements.js/tree/babel-testcase
2. Run `yarn dlx gulp` (where the yarn version is 3.x)

The first time you run gulp you'll get one error. The second time you run, you'll get the aforementioned error about `convert-source-map`.

Please confirm whether you are able to reproduce the problem now.

**Comment by nicolo-ribaudo** [comment]
I see a different error, but probably it's related

```
You must provide the URL of lib/mappings.wasm by calling SourceMapConsumer.initialize({ 'lib/mappings.wasm': ... }) before using SourceMapConsumer
```

**Comment by nicolo-ribaudo** [comment]
Ok the error above is because the used `source-map` version used by `gulp-typescript` assumes that if there is a global `fetch`, then we are in a browser: https://unpkg.com/browse/source-map@0.7.3/lib/read-wasm.js. If I disable the check in that line (which is unrelated from Babel), then the build passes successfully and I cannot reproduce the error caused by `@babel/register` :thinking:

**Comment by cowwoc** [comment]
Based on what I saw, this problem only occurs if the previous build did not completely successfully.

Meaning, if the first run failed due to a missing dependency then the second run would have the aforementioned error. But if the build succeeded, then cached runs would work fine.

This seems to match the behavior you saw (the exception would only get thrown if `source-map` triggered a build error).

At this point I got my build working, so this issue no longer blocks me. I'll leave it up to you whether you want to investigate further or close the issue. Thank you for your investigate either way.
